lms/static/js/learner_dashboard/views/course_entitlement_view.js

```diff
@@ -0,0 +1,309 @@
+                     2) Add the new session's dates to the date field on the main course card.
+                     3) Hide the 'View Course' button to the course card.
+
+                     On unenroll:
+                     1) Hide the change session button and the date field.
+                     2) Hide the 'View Course' button.
+                     3) Remove the messages associated with the enrolled state.
+                     4) Remove the link from the course card image and title.
+                     */
+                     var enrolled = data.is_active;
+
+                     // Update the model with the new session Id;
+                     this.entitlementModel.set({currentSessionId: this.currentSessionSelection});
+                     if (enrolled) {
+                         this.$triggerOpenBtn.removeClass('hidden');
+                         this.$dateDisplayField.html(this.getAvailableSessionWithId(data.course_details.course_id)
+                             .session_dates).prepend('<span class="fa fa-check" aria-hidden="true"></span>');
+                         this.$enterCourseBtn.attr('href', this.formatCourseHomeUrl(data.course_details.course_id))
+                             .removeClass('hidden');
+                         this.toggleSessionSelectionPanel();
+                     } else {
+                         // Reset the card contents to the unenrolled state
+                         this.$triggerOpenBtn.addClass('hidden');
+                         this.$enterCourseBtn.addClass('hidden');
+                         this.$dateDisplayField.html('');
+                         this.$courseCardMessages.remove();
+
+                         // Remove links to previously enrolled sessions
+                         this.$courseImageLink.replaceWith('<div class="' + this.$courseImageLink.attr('class') +
+                             '" tabindex="-1">' + this.$courseImageLink.html() + '</div>');
+                         this.$courseTitleLink.replaceWith('<span>' + this.$courseTitleLink.text() + '</span>');
+                     }
+                 },
+
+                 enrollError: function() {
+                     this.$dateDisplayField.find('.fa.fa-spin').removeClass('fa-spin fa-spinner').addClass('fa-close');
+                     this.$dateDisplayField
+                         .append(gettext('There was an error, please reload the page and try again.'));
+                     this.hideVerificationDialog();
+                 },
+
+                 updateEnrollBtn: function() {
+                     /*
+                     This function is invoked on load, on opening the view and on changing the option on the session
+                     selection dropdown. It plays three roles:
+                     1) Enables and disables enroll button
+                     2) Changes text to describe the action taken
+                     3) Formats the confirmation popover to allow for two step authentication
```

Comment:
Where is ```popover``` coming from? I assume Bootstrap? It would be great to document this in the code to provide context.

Sometimes it is common to wrap this sort of vendor-specific functionality in non-specific functions, e.g. 
```
closeTooltip($el) {
    $el.popover('dispose');
}
this.closeTooltip(enrollBtnInitial);
```
This often make it clearer what is going on in the code and makes it easier to replace the vendor package if we want to later down the line (including upgrading versions of Bootstrap). It seems like you would only really need 3 wrapper functions, to;

1. Create a new tooltip
2. Open a tooltip
3. Close a tooltip

So I think it could be worthwhile to do, and then you would just add comments above the wrapper functions noting that the function uses Bootstrap's popover.

Reply:
I broke it out into a new template and helper functions that make it clear what is happening.

Reply:
It's a good start adding the ```hideVerificationDialog``` and ```removeVerificationDialog``` functions, but it would be great to add move generic ```hideDialog``` and ```showDialog``` functions that accept an element to wrap all of the vendor-specific code.